ethcore/src/client/traits.rs

```diff
@@ -194,15 +194,20 @@ pub trait BlockChainClient : Sync + Send {
 	fn gas_price_statistics(&self, sample_size: usize, distribution_size: usize) -> Result<Vec<U256>, ()> {
 		let mut h = self.chain_info().best_block_hash;
 		let mut corpus = Vec::new();
-		for _ in 0..sample_size {
-			let block_bytes = self.block(BlockID::Hash(h)).expect("h is either the best_block_hash or an ancestor; qed");
-			let block = BlockView::new(&block_bytes);
-			let header = block.header_view();
-			if header.number() == 0 {
-				break;
+		while corpus.is_empty() {
+			for _ in 0..sample_size {
+				let block_bytes = self.block(BlockID::Hash(h)).expect("h is either the best_block_hash or an ancestor; qed");
```

Comment:
not sure how much longer this invariant will be upheld, given that pruning history is often < 100.